# SmartPredict — Data Understanding

## AI4I 2020 Predictive Maintenance Dataset

The objective of this project is to develop an end-to-end machine learning
system capable of predicting industrial machine failures using operational
and sensor-related features.

This notebook focuses on understanding the raw dataset before preprocessing
and model development.

In [1]:
import pandas as pd 
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

from pathlib import Path
from ucimlrepo import fetch_ucirepo

In [2]:
dataset = fetch_ucirepo(id=601)

print(dataset.metadata.name)

AI4I 2020 Predictive Maintenance Dataset


In [3]:
df = dataset.data.original.copy()

df.head()

,UID,Product ID,Type,Air temperature,Process temperature,Rotational speed,Torque,Tool wear,Machine failure,TWF,HDF,PWF,OSF,RNF
0,1,M14860,M,298.1,308.6,1551,42.8,0,0,0,0,0,0,0
1,2,L47181,L,298.2,308.7,1408,46.3,3,0,0,0,0,0,0
2,3,L47182,L,298.1,308.5,1498,49.4,5,0,0,0,0,0,0
3,4,L47183,L,298.2,308.6,1433,39.5,7,0,0,0,0,0,0
4,5,L47184,L,298.2,308.7,1408,40.0,9,0,0,0,0,0,0


In [4]:
raw_data_path = Path("../data/raw/ai4i2020.csv")

df.to_csv(raw_data_path, index=False)

print(f"Dataset saved to: {raw_data_path}")

Dataset saved to: ..\data\raw\ai4i2020.csv


In [5]:
print(f"Rows: {df.shape[0]}")
print(f"Columns: {df.shape[1]}")

Rows: 10000
Columns: 14


In [6]:
df.columns.tolist()

['UID',
 'Product ID',
 'Type',
 'Air temperature',
 'Process temperature',
 'Rotational speed',
 'Torque',
 'Tool wear',
 'Machine failure',
 'TWF',
 'HDF',
 'PWF',
 'OSF',
 'RNF']

In [7]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   UID                  10000 non-null  int64  
 1   Product ID           10000 non-null  str    
 2   Type                 10000 non-null  str    
 3   Air temperature      10000 non-null  float64
 4   Process temperature  10000 non-null  float64
 5   Rotational speed     10000 non-null  int64  
 6   Torque               10000 non-null  float64
 7   Tool wear            10000 non-null  int64  
 8   Machine failure      10000 non-null  int64  
 9   TWF                  10000 non-null  int64  
 10  HDF                  10000 non-null  int64  
 11  PWF                  10000 non-null  int64  
 12  OSF                  10000 non-null  int64  
 13  RNF                  10000 non-null  int64  
dtypes: float64(3), int64(9), str(2)
memory usage: 1.1 MB


In [8]:
df.describe()

,UID,Air temperature,Process temperature,Rotational speed,Torque,Tool wear,Machine failure,TWF,HDF,PWF,OSF,RNF
count,10000.00000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.00000
mean,5000.50000,300.004930,310.005560,1538.776100,39.986910,107.951000,0.033900,0.004600,0.011500,0.009500,0.009800,0.00190
std,2886.89568,2.000259,1.483734,179.284096,9.968934,63.654147,0.180981,0.067671,0.106625,0.097009,0.098514,0.04355
min,1.00000,295.300000,305.700000,1168.000000,3.800000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.00000
25%,2500.75000,298.300000,308.800000,1423.000000,33.200000,53.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.00000
50%,5000.50000,300.100000,310.100000,1503.000000,40.100000,108.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.00000
75%,7500.25000,301.500000,311.100000,1612.000000,46.800000,162.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.00000
max,10000.00000,304.500000,313.800000,2886.000000,76.600000,253.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.00000


In [15]:
df.describe(include="str")

,Product ID,Type
count,10000,10000
unique,10000,3
top,M14860,L
freq,1,6000


In [10]:
df.isnull().sum()

UID                    0
Product ID             0
Type                   0
Air temperature        0
Process temperature    0
Rotational speed       0
Torque                 0
Tool wear              0
Machine failure        0
TWF                    0
HDF                    0
PWF                    0
OSF                    0
RNF                    0
dtype: int64

In [11]:
print("Duplicate rows:", df.duplicated().sum())

Duplicate rows: 0


In [12]:
df["Machine failure"].value_counts()

Machine failure
0    9661
1     339
Name: count, dtype: int64

In [13]:
df["Machine failure"].value_counts(normalize=True) * 100

Machine failure
0    96.61
1     3.39
Name: proportion, dtype: float64

In [14]:
failure_columns = [
    "TWF",
    "HDF",
    "PWF",
    "OSF",
    "RNF"
]

df[failure_columns].sum().sort_values(ascending=False)

HDF    115
OSF     98
PWF     95
TWF     46
RNF     19
dtype: int64

## Product Type Distribution

In [16]:
df["Type"].value_counts()

Type
L    6000
M    2997
H    1003
Name: count, dtype: int64

In [17]:
df["Type"].value_counts(normalize=True) * 100

Type
L    60.00
M    29.97
H    10.03
Name: proportion, dtype: float64

In [18]:
df.groupby("Machine failure")[["TWF", "HDF", "PWF", "OSF", "RNF"]].sum()

,TWF,HDF,PWF,OSF,RNF
Machine failure,,,,,
0,0,0,0,0,18
1,46,115,95,98,1


In [19]:
df["failure_count"] = df[
    ["TWF", "HDF", "PWF", "OSF", "RNF"]
].sum(axis=1)

df["failure_count"].value_counts().sort_index()

failure_count
0    9652
1     324
2      23
3       1
Name: count, dtype: int64

In [20]:
df[df["failure_count"] > 1][
    [
        "Machine failure",
        "TWF",
        "HDF",
        "PWF",
        "OSF",
        "RNF"
    ]
].head(10)

,Machine failure,TWF,HDF,PWF,OSF,RNF
69,1,0,0,1,1,0
1324,1,0,0,1,1,0
1496,1,0,0,1,1,0
3611,1,1,0,0,0,1
3854,1,0,0,1,1,0
3943,1,0,0,1,1,0
4254,1,0,1,1,0,0
4342,1,0,1,1,0,0
4370,1,0,1,0,1,0
4383,1,0,1,0,1,0


In [21]:
failure_flags = ["TWF", "HDF", "PWF", "OSF", "RNF"]

derived_failure = (df[failure_flags].sum(axis=1) > 0).astype(int)

pd.crosstab(
    df["Machine failure"],
    derived_failure,
    rownames=["Actual Machine Failure"],
    colnames=["Derived From Failure Flags"]
)

Derived From Failure Flags,0,1
Actual Machine Failure,,
0,9643,18
1,9,330


In [22]:
df.drop(columns=["failure_count"], inplace=True)

In [23]:
df.shape

(10000, 14)

## Initial Findings

- The dataset contains 10,000 observations and 14 variables.
- No missing values were detected.
- No duplicated rows were detected.
- Machine failures represent only 3.39% of the dataset, indicating a significant class imbalance.
- Product types are distributed as L (60.00%), M (29.97%), and H (10.03%).
- UID and Product ID are identifiers and will not be used as predictive features.
- TWF, HDF, PWF, OSF and RNF directly describe failure modes and will be excluded from model inputs to prevent target leakage.
- A small inconsistency exists between the overall machine-failure label and individual failure-mode indicators.
- Product Type is categorical and will require encoding.
- Evaluation metrics such as Recall, Precision, F1-score and PR-AUC will be more informative than accuracy alone.

### Target Consistency Check

A consistency check was performed between the `Machine failure` target and
the individual failure indicators (`TWF`, `HDF`, `PWF`, `OSF`, `RNF`).

Results showed that:

- 330 machine failure observations contain at least one explicit failure flag.
- 18 observations contain a failure flag while `Machine failure` is 0.
- 9 observations are labeled as machine failures although none of the five
  failure indicators are active.

These observations indicate a small amount of label inconsistency or noise
in the dataset. The original `Machine failure` column will be retained as
the prediction target.